src/ring_theory/local_properties.lean

```diff
@@ -0,0 +1,125 @@
+/-- An ideal is trivial if its localization at every maximal ideal is trivial. -/
+lemma ideal_eq_zero_of_localization (I : ideal R)
+   (h : ∀ (J : ideal R) (hJ : J.is_maximal),
+      by exactI is_localization.coe_submodule (localization.at_prime J) I = 0) : I = 0 :=
+begin
+  by_contradiction hI,
+  obtain ⟨x, hx, hx'⟩ := set.exists_of_ssubset (bot_lt_iff_ne_bot.mpr hI),
+  rw [submodule.bot_coe, set.mem_singleton_iff] at hx',
+  have H : (ideal.span ({x} : set R)).annihilator ≠ ⊤,
+  { rw [ne.def, submodule.annihilator_eq_top_iff],
+    by_contra,
+    apply hx',
+    rw [← set.mem_singleton_iff, ← @submodule.bot_coe R, ← h],
+    exact ideal.subset_span (set.mem_singleton x) },
+  obtain ⟨p, hp₁, hp₂⟩ := ideal.exists_le_maximal _ H,
+  resetI,
+  specialize h p hp₁,
+  have : algebra_map R (localization.at_prime p) x = 0,
+  { rw ← set.mem_singleton_iff,
+    change algebra_map R (localization.at_prime p) x ∈ (0 : submodule R (localization.at_prime p)),
+    rw ← h,
+    exact submodule.mem_map_of_mem hx },
+  rw is_localization.map_eq_zero_iff p.prime_compl at this,
+  obtain ⟨m, hm⟩ := this,
+  apply m.prop,
+  refine hp₂ _,
+  erw submodule.mem_annihilator_span_singleton,
+  rwa mul_comm at hm,
+end
+
+lemma eq_zero_of_localization (r : R)
+   (h : ∀ (J : ideal R) (hJ : J.is_maximal),
+      by exactI algebra_map R (localization.at_prime J) r = 0) : r = 0 :=
+begin
+  rw ← ideal.span_singleton_eq_bot,
+  apply ideal_eq_zero_of_localization,
+  intros J hJ,
+  delta is_localization.coe_submodule,
+  erw [submodule.map_span, submodule.span_eq_bot],
+  rintro _ ⟨_, h', rfl⟩,
+  cases set.mem_singleton_iff.mpr h',
+  exact h J hJ,
+end
+
+end ideal
+
+section reduced
+
+lemma localization_is_reduced [is_localization M R'] [hR : is_reduced R] : is_reduced R' :=
+begin
+  constructor,
+  rintro x ⟨(_|n), e⟩,
+  { simpa using congr_arg (*x) e },
+  obtain ⟨⟨y, m⟩, hx⟩ := is_localization.surj M x,
+  dsimp only at hx,
+  let hx' := congr_arg (^ n.succ) hx,
+  simp only [mul_pow, e, zero_mul, ← ring_hom.map_pow] at hx',
+  rw [← (algebra_map R R').map_zero] at hx',
+  obtain ⟨m', hm'⟩ := (is_localization.eq_iff_exists M R').mp hx',
+  apply_fun (*m'^n) at hm',
+  simp only [mul_assoc, zero_mul] at hm',
+  rw [mul_comm, ← pow_succ, ← mul_pow] at hm',
+  replace hm' := is_nilpotent.eq_zero ⟨_, hm'.symm⟩,
+  rw [← (is_localization.map_units R' m).mul_left_inj, hx, zero_mul,
+    is_localization.map_eq_zero_iff M],
+  exact ⟨m', by rw [← hm', mul_comm]⟩
+end
+
+instance [is_reduced R] : is_reduced (localization M) := localization_is_reduced M _
+
+lemma is_reduced_of_localization (h : ∀ (J : ideal R) (hJ : J.is_maximal),
+  by exactI is_reduced (localization.at_prime J)) : is_reduced R :=
+begin
+  constructor,
+  intros x hx,
+  apply eq_zero_of_localization,
+  intros J hJ,
+  specialize h J hJ,
+  resetI,
+  exact (hx.map (algebra_map R (localization.at_prime J)).to_monoid_with_zero_hom).eq_zero,
```

Comment:
With the change to `is_nilpotent.map` you should be able to write:
```suggestion
  exact (hx.map (algebra_map R (localization.at_prime J))).eq_zero,
```
(can't test it yet because the code is still compiling)